### PR TITLE
feat(anta): Added the test case to verify Update error counters for BGP neighbors

### LINKED ISSUE #770: Add the test case to verify Update error counters for BGP neighbors

### PR DESCRIPTION
**Pull request by vitthalmagadum**
# Description

Verifies the BGP peer's errors counters should be zero.

    Expected Results
    ----------------
    * Success: The test will pass if the BGP peer's error counters are zero.
    * Failure: The test will fail if the BGP peer's error counters are non-zero.

Fixes #770 

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have run pre-commit for code linting and typing (`pre-commit run`)
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

File: anta/custom_types.py [hunk 1]
```diff
@@ -167,3 +167,4 @@ def validate_regex(value: str) -> str:
 Hostname = Annotated[str, Field(pattern=REGEXP_TYPE_HOSTNAME)]
 Port = Annotated[int, Field(ge=1, le=65535)]
 RegexString = Annotated[str, AfterValidator(validate_regex)]
+BgpErrors = Literal["inUpdErrWithdraw", "inUpdErrIgnore", "inUpdErrDisableAfiSafi", "disabledAfiSafi", "lastUpdErrTime"]
```
Comment (carl-baillargeon) [inline anta/custom_types.py:170]:
```suggestion
BgpUpdateErrors = Literal["inUpdErrWithdraw", "inUpdErrIgnore", "inUpdErrDisableAfiSafi", "disabledAfiSafi", "lastUpdErrTime"]
```

File: anta/tests/routing/bgp.py [hunk 2]
```diff
@@ -1226,3 +1226,81 @@ def test(self) -> None:
             self.result.is_success()
         else:
             self.result.is_failure(f"Following BGP peers are not configured or hold and keep-alive timers are not correct:\n{failures}")
+
+
+class VerifyBGPPeerErrors(AntaTest):
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1231]:
```suggestion
class VerifyBGPPeerUpdateErrors(AntaTest):
```

File: anta/tests/routing/bgp.py [hunk 3]
```diff
@@ -1226,3 +1226,81 @@ def test(self) -> None:
...
+        """List of BGP peers"""
+
+        class BgpPeer(BaseModel):
+            """Model for a BGP peer."""
+
+            peer_address: IPv4Address
+            """IPv4 address of a BGP peer."""
+            vrf: str = "default"
+            """Optional VRF for BGP peer. If not provided, it defaults to `default`."""
+            errors: list[BgpErrors]
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1271]: Like #778, we should have something like:
```
update_error_filter: list[BgpUpdateError] | None = None
```
None means we check all the errors, other we check only the provided ones.

File: anta/tests/routing/bgp.py [hunk 4]
```diff
@@ -1226,3 +1226,81 @@ def test(self) -> None:
...
+            """List of error counters to be verified."""
+
+    def render(self, template: AntaTemplate) -> list[AntaCommand]:
+        """Render the template for each Bgp peer in the input list."""
+        return [template.render(peer=str(bgp_peer.peer_address), vrf=bgp_peer.vrf) for bgp_peer in self.inputs.bgp_peers]
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifyEVPNType2Route."""
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1280]:
```suggestion
        """Main test function for VerifyBGPPeerUpdateErrors."""
```

File: anta/tests/routing/bgp.py [hunk 5]
```diff
@@ -1226,3 +1226,81 @@ def test(self) -> None:
...
+        failures: dict[Any, Any] = {}
+
+        for command, input_entry in zip(self.instance_commands, self.inputs.bgp_peers):
+            peer = command.params.peer
+            vrf = command.params.vrf
+            error_counters = input_entry.errors
+
+            # Verify BGP peer
+            if not (bgp_output := get_value(command.json_output, f"vrfs.{vrf}.peerList")) or (bgp_output := get_item(bgp_output, "peerAddress", peer)) is None:
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1289]: Like my comment in #778, please change the variable names for clarity.

File: anta/tests/routing/bgp.py [hunk 6]
```diff
@@ -1226,3 +1226,81 @@ def test(self) -> None:
...
+                for error_conter in error_counters
+            ]
+            if any(error_counters_not_ok):
+                failures[peer] = {vrf: error_counters_output}
+
+        # Check if any failures
+        if not failures:
+            self.result.is_success()
+        else:
+            self.result.is_failure(f"Following BGP peers are not configured or error counters are not correct:\n{failures}")
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1306]: Similar comment in #778

File: anta/tests/routing/bgp.py [hunk 7]
```diff
@@ -1226,3 +1226,81 @@ def test(self) -> None:
...
+
+            # Verify BGP peer
+            if not (bgp_output := get_value(command.json_output, f"vrfs.{vrf}.peerList")) or (bgp_output := get_item(bgp_output, "peerAddress", peer)) is None:
+                failures[peer] = {vrf: "Not configured"}
+                continue
+
+            # Verify BGP peer's error counters
+            error_counters_output = bgp_output.get("peerInUpdateErrors", {})
+            error_counters_not_ok = [
+                None if error_conter == "disabledAfiSafi" and error_counters_output.get(error_conter) == "None" else error_counters_output.get(error_conter)
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1296]: Typo here on `error_conter`. That logic will probably change anyways when you refactor with `update_error_filter` input option.

File: anta/tests/routing/bgp.py (hunk 2 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Handled

File: anta/tests/routing/bgp.py (hunk 3 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Added check for verifying all the error counters and if any mentioned error counter not found.

File: anta/tests/routing/bgp.py (hunk 4 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Handled

File: anta/tests/routing/bgp.py (hunk 5 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Handled

File: anta/tests/routing/bgp.py (hunk 6 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Handled

File: anta/tests/routing/bgp.py (hunk 7 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Handled

File: anta/custom_types.py [hunk 8]
```diff
@@ -167,3 +167,4 @@ def validate_regex(value: str) -> str:
 Hostname = Annotated[str, Field(pattern=REGEXP_TYPE_HOSTNAME)]
 Port = Annotated[int, Field(ge=1, le=65535)]
 RegexString = Annotated[str, AfterValidator(validate_regex)]
+BgpUpdateErrors = Literal["inUpdErrWithdraw", "inUpdErrIgnore", "inUpdErrDisableAfiSafi", "disabledAfiSafi", "lastUpdErrTime"]
```
Comment (carl-baillargeon) [inline anta/custom_types.py:170]: Let's use singular here since the type is basically an individual error and not multiple types of errors.
```suggestion
BgpUpdateError = Literal["inUpdErrWithdraw", "inUpdErrIgnore", "inUpdErrDisableAfiSafi", "disabledAfiSafi", "lastUpdErrTime"]
```

File: anta/tests/routing/bgp.py [hunk 9]
```diff
@@ -1226,3 +1226,88 @@ def test(self) -> None:
...
             self.result.is_failure(f"Following BGP peers are not configured or hold and keep-alive timers are not correct:\n{failures}")
+
+
+class VerifyBGPPeerUpdateErrors(AntaTest):
+    """Verifies if the IPv4 BGP peer's update errors counters should be zero.
+
+    Expected Results
+    ----------------
+    * Success: The test will pass if the BGP peer's update error counters are zero.
+    * Failure: The test will fail if the BGP peer's update error counters are non-zero.
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1237]: Please update this following the suggestion I made in #778 for the docstring.

File: anta/tests/routing/bgp.py [hunk 10]
```diff
@@ -1226,3 +1226,88 @@ def test(self) -> None:
...
+
+            peer_address: IPv4Address
+            """IPv4 address of a BGP peer."""
+            vrf: str = "default"
+            """Optional VRF for BGP peer. If not provided, it defaults to `default`."""
+            update_error_filter: list[BgpUpdateErrors] | None = None
+            """Optional list of update error counters to be verified. If not provided, test will verifies all the update error counters."""
+
+    def render(self, template: AntaTemplate) -> list[AntaCommand]:
+        """Render the template for each Bgp peer in the input list."""
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1275]:
```suggestion
        """Render the template for each BGP peer in the input list."""
```

File: tests/units/anta_tests/routing/test_bgp.py [hunk 11]
```diff
@@ -3722,4 +3723,322 @@
...
+                "vrfs": {
+                    "default": {
+                        "peerList": [
+                            {
+                                "peerAddress": "10.100.0.8",
+                                "peerInUpdateErrors": {
+                                    "inUpdErrWithdraw": 0,
+                                    "inUpdErrIgnore": 0,
+                                    "inUpdErrDisableAfiSafi": 0,
+                                    "disabledAfiSafi": "Yes",
```
Comment (carl-baillargeon) [inline tests/units/anta_tests/routing/test_bgp.py:3810]: Did you validate that "Yes" is used when it's not "None"?

File: anta/tests/routing/bgp.py [hunk 12]
```diff
@@ -1226,3 +1226,88 @@ def test(self) -> None:
...
+
+            # In case update error counters not provided, It will check all the update error counters.
+            if not update_error_counters:
+                update_error_counters = error_counters_output
+
+            # verifying the error counters.
+            error_counters_not_ok = {
+                ("disabledAfiSafi" if error_counter == "disabledAfiSafi" else error_counter): value
+                for error_counter in update_error_counters
+                if (value := error_counters_output.get(error_counter, "Not Found")) != "None" and value != 0
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1304]: I wonder if we could have other values than "None" here that could also represent a bad state. Like "False" or something...

File: tests/units/anta_tests/routing/test_bgp.py (hunk 11 above)
Comment (carl-baillargeon) [reply tests/units/anta_tests/routing/test_bgp.py:3810]: I think it would be something like "ipv4Unicast" or something.

File: anta/tests/routing/bgp.py (hunk 12 above)
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1304]: Let's add a note in the docstring of the class for this "None" corner case saying that for the `disabledAfiSafi` field we check that it's not "None' versus 0.

File: anta/tests/routing/bgp.py [hunk 13]
```diff
@@ -1226,3 +1226,88 @@ def test(self) -> None:
...
+                vrf: default
+                errors:
+                  - inUpdErrWithdraw
+    ```
+    """
+
+    name = "VerifyBGPPeerUpdateErrors"
+    description = "Verifies the update error counters of a BGP IPv4 peer."
+    categories: ClassVar[list[str]] = ["bgp"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaTemplate(template="show bgp neighbors {peer} vrf {vrf}")]
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:1256]: Please add the revision number for the command. Thanks

Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Added most current revision 3.

File: anta/tests/routing/bgp.py (hunk 12 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py:1310]: Added Note in docstring for the same.

File: tests/units/anta_tests/routing/test_bgp.py (hunk 11 above)
Comment (vitthalmagadum) [reply tests/units/anta_tests/routing/test_bgp.py]: Thanks for the catching up.
I have just added a string for checking. Forgot to remove. Thanks!!

File: anta/tests/routing/bgp.py (hunk 10 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Handled.

File: anta/tests/routing/bgp.py (hunk 9 above)
Comment (vitthalmagadum) [reply anta/tests/routing/bgp.py]: Updated the docstring.

File: anta/custom_types.py (hunk 8 above)
Comment (vitthalmagadum) [reply anta/custom_types.py]: Updated

**Review by carl-baillargeon — APPROVED** [review]
LGTM
